mobify/progressive-web-scaffold
# [Payment] Add form submission functionality

**JIRA**: https://mobify.atlassian.net/browse/WEB-1112

## Changes
- Use checkout selectors to grab shipping address information
- Add payment form submission functionality
- Set initial value for 'same as shipping address' to `true` in the parser
- Add form selector to grab payment billing address form
- Show actual saved address content under the checkbox field
- Save payment address to state so in confirmation page, we have access to the address data
- Add `fetchUrl` prop to route because `/checkout/payment/` is not a valid URL on desktop so we wouldn't be able to process and grab initial values.
  - We also tried `<Route component={CheckoutPayment} path="checkout/#payment" routeName="checkout-payment" Header={CheckoutHeader} Footer={CheckoutFooter} />`, it works fine when we go from Shipping -> Payment. But when we refresh on the payment page, the route matches to the Shipping container for some reason.

## TODO
- Need to grab 'username' (email) from shipping step. Jen will make that data persist to payment page, so leaving out this feature for now. Payment form requires email data for form submission as well. (I'm hardcoding it as `mobifyqa@gmail.com`) for now.
- Have yet to figure out what to do if the user refreshes on the Payment page. It kinda breaks at the moment. We deemed this is pretty edge case and since we are tight on time, we'll revisit fixing this at a later time.

## Know issues
- Confirmation page is currently broken so we won't actually see the confirmation page show up. But in the console you should see an error when trying to render the confirmation page.
- (Guest checkout only) Cart count doesn't reset to 0 when order is successfully place. However if you go `https://www.merlinspotions.com/checkout/` you'll be redirected to the cart page and the cart count will get updated to 0 correctly. After some investigation with Jen, we feel this bug will be fixed with the confirmation page work

## How to test-drive this PR
- Run `npm run dev`
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fwww.merlinspotions.com%2F&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1)
- add an item to the cart
- navigate to: https://www.merlinspotions.com/checkout/
- Fill in form, proceed to payment step
- Either leave same address checkbox checked or add new address, submit form
- You should now end up on confirmation page
